cloudpathlib/s3/s3client.py

```diff
@@ -57,18 +62,34 @@ def __init__(
             boto3_transfer_config (Optional[dict]): Instantiated TransferConfig for managing s3 transfers.
                 (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/s3.html#boto3.s3.transfer.TransferConfig)
         """
-        if boto3_session is not None:
-            self.sess = boto3_session
-        else:
-            self.sess = Session(
-                aws_access_key_id=aws_access_key_id,
-                aws_secret_access_key=aws_secret_access_key,
-                aws_session_token=aws_session_token,
-                botocore_session=botocore_session,
-                profile_name=profile_name,
```

Comment:
I think you actually want to keep the block of logic related to the session up here on the outside of `if no_sign_request`. 
- If `boto3_session` is passed, it should take precedence. 
- Otherwise it uses the arguments, which all default to `None`, which would be the same as what you have with `self.sess = Session()` in your `if no_sign_request` branch. 

```
if boto3_session is not None:
    # set passed session
else:
    # create session from args

if no_sign_request:
    # create s3 and client with UNSIGNED config
else:
    # create s3 and client without config
```

Reply:
Yeah that is how I started implementing it, but when looking at the AWS CLI docs they say the creds won't even be loaded if `--no-sign-request` is passed to the CLI:
https://awscli.amazonaws.com/v2/documentation/api/latest/reference/index.html

As implemented, we basically mimic that (ignore all creds and don't even load them if `no_sign_request`). Agreed that mixing and matching `no_sign_request` with any of the other options could potentially cause confusion.

There's not enough about UNSIGNED in the boto3 docs to track down how that config works with creds being passed as well so not sure what the implications of creating the session with user passed creds then not signing are.

Reply:
Both versions pass our test so will keep the session creds around if they are passed.

Reply:
I think the test should be enough, right? 

---

Went to look at what awscli uses, and they go through a slightly different mechanism in botocore I guess.

https://github.com/aws/aws-cli/blob/12956b24b88e0b60f6a871013d153f4e513f030f/awscli/customizations/globalargs.py#L87-L90

https://github.com/boto/botocore/blob/c75af12b09618a37b6c7f0bfac5f8cc6aa0bb7e7/botocore/handlers.py#L284-L289

Reply:
Yeah, we got this approach from their issues, so should be reliable:
https://github.com/boto/boto3/issues/1200

I tested both with and without a `~/.aws/credentials` file and confirmed that I could access public S3 items without any credentials when I passed the param

Reply:
In any case, I don't think your implementation actually adds any benefit in terms of not loading any credentials. If you initiate `Session()`, it just initializes with a default of `None` for everything, which will load from env vars.

Reply:
🤷 maybe. I didn't dig deep enough in Boto3 to track when they do the actual loading and see if they pay attention to `UNSIGNED` when determining if they should even look for credentials.